File: src/js/lattice/unit_cell.ts
import { Vector3DSchema } from "@mat3ra/esse/dist/js/types";

export type UnitCellProps = [
    number,
    number,
    number,
    number,
    number,
    number,
    number,
    number,
    number,
    string,
];

// TODO: refactor and remove the need for UnitCell
export class UnitCell {
    ax: number;

    ay: number;

    az: number;

    bx: number;

    by: number;

    bz: number;

    cx: number;

    cy: number;

    cz: number;

    units: string;

    constructor([ax, ay, az, bx, by, bz, cx, cy, cz, units]: UnitCellProps) {
        this.ax = ax;
        this.ay = ay;
        this.az = az;
        this.bx = bx;
        this.by = by;
        this.bz = bz;
        this.cx = cx;
        this.cy = cy;
        this.cz = cz;
        this.units = units;
    }

    vectorA(): Vector3DSchema {
        return [this.ax, this.ay, this.az];
    }

    vectorB(): Vector3DSchema {
        return [this.bx, this.by, this.bz];
    }

    vectorC(): Vector3DSchema {
        return [this.cx, this.cy, this.cz];
    }

    axes(): [Vector3DSchema, Vector3DSchema, Vector3DSchema] {
        return [this.vectorA(), this.vectorB(), this.vectorC()];
    }
}

File: src/py/mat3ra/made/lattice/unit_cell.py
from typing import List


class UnitCell:
    ax: float
    ay: float
    az: float
    bx: float
    by: float
    bz: float
    cx: float
    cy: float
    cz: float
    units: str

    def __init__(self, vectors):
        self.ax, self.ay, self.az, self.bx, self.by, self.bz, self.cx, self.cy, self.cz, self.units = vectors

    @property
    def vectorA(self) -> List[float]:
        return [self.ax, self.ay, self.az]

    @property
    def vectorB(self) -> List[float]:
        return [self.bx, self.by, self.bz]

    @property
    def vectorC(self) -> List[float]:
        return [self.cx, self.cy, self.cz]

    @property
    def axes(self) -> List[List[float]]:
        return [self.vectorA, self.vectorB, self.vectorC]
